h5py/_hl/dataset.py

```diff
@@ -67,7 +67,11 @@ def make_new_dset(parent, shape=None, dtype=None, data=None,
     # Convert data to a C-contiguous ndarray
     if data is not None and not isinstance(data, Empty):
         from . import base
-        data = numpy.asarray(data, order="C", dtype=base.guess_dtype(data))
+        if dtype is None:
+            _dtype = base.guess_dtype(data)
+        else:
+            _dtype = dtype
+        data = numpy.asarray(data, order="C", dtype=_dtype)
```

Comment:
This change looks valid. There is no point in converting to an intermediate type if the user has given a specific type as keyword.

The code could be shortened as `data = numpy.asarray(data, order="C", dtype=dtype or base.guess_dtype(data))`